### ISSUE TITLE
[array-callback-return] False positive when in a map that uses a switch case without default because all possibilities are handled but ESLint isn't aware of that

### ISSUE THREAD

**Issue by MaximeCheramy** [typescript-eslint/typescript-eslint#2841]
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

[playground link](https://typescript-eslint.io/play/#ts=5.3.3&fileType=.tsx&code=MYewdgzgLgBANjAvDA2gbxlAngBwKYBcMA5AIbEwC%2BANDBtvkcQEYWUC6MpEdmuhJcjAA%2BJVlRTsAUKEiwAZiBBJ4AOgC2pHAAo8CRAD46UmDAgB3AJZRgACxi64qhngCUx06eDc8g4kQAnPCgAVwCwEhcuYhNPbwhfFn8YINDwyP4YVljKKUpXIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQIbWlwE8BaAY13ngCNcyBrE6RAF1mgDt0oB3fL8GAC%2BIIUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUhFYASSSomyPACV0eNplIUwo9FIC%2BIQ0A&tokens=false)

The rule `array-callback-return` is enabled with default settings. There are no TypeScript equivalent (that's what would fix the issue).

```TS
    const l = [{ type: 'a' }, { type: 'b' }] as {type: 'a' | 'b'}[]
    l.map(el => {
      switch (el.type) {
        case 'a': return 'type a'
        case 'b': return 'type b'
      }
    })
```

```json
{
  "compilerOptions": {
    "declaration": true,
    "declarationMap": true,
    "lib": [
      "dom",
      "es2018",
      "es2018.regexp"
    ],
    "module": "es6",
    "moduleResolution": "node",
    "outDir": "dist",
    "sourceMap": true,
    "target": "es5",
    "types": [
      "jest",
      "node"
    ]
  },
  "include": [
    "src/**/*",
    "tests/**/*"
  ]
}

```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.9.0` |
| `@typescript-eslint/parser`        | `4.9.0` |
| `TypeScript`                       | `4.1.2` |
| `ESLint`                           | `7.14.0` |
| `node`                             | `12.19.0` |

**Comment by nicolassanmar** [comment]
As a workaround while we don't get this implemented as a TS rule, we can do the following to ensure we both always return a value and also did not miss any case in the `switch` statement , without having to disable `array-callback-return` or `no-unreachable`:

``` ts
const exhaustivenessSafetyCheck = (never: never) => {
  throw new Error(`Unknown value ${never}`)
}

const type: 'A' | 'B'
// Example usage
 switch (type) {
  case 'A':
    return ...
  case 'B':
    return ...
  default:
    return exhaustivenessSafetyCheck(type)
  }
```

This only accepts `never` as a type (triggering type errors if you pass anything else) and throws an error when called. If we forget to check one case in the `switch`, TS will complain that the argument passed to this function is not of type never

**Comment by kirkwaiblinger** [comment]
I think we could close this as being better handled by TS's [noImplicitReturns](https://www.typescriptlang.org/tsconfig/#noImplicitReturns) option. ([playground link with this enabled](https://typescript-eslint.io/play/#ts=5.3.3&fileType=.tsx&code=MYewdgzgLgBANjAvDA2gbxlAngBwKYBcMA5AIbEwC%2BANDBtvkcQEYWUC6MpEdmuhJcjAA%2BJVlRTsAUKEiwAZiBBJ4AOgC2pHAAo8CRAD46UmDAgB3AJZRgACxi64qhngCUx06eDc8g4kQAnPCgAVwCwEhcuYhMqKUpXIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQIbWlwE8BaAY13ngCNcyBrE6RAF1mgDt0oB3fL8GAC%2BIIUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUhFYASSSomyPACV0eNplIUwo9FIC%2BIQ0A&tokens=false))

AFAICT the only case it doesn't cover is when there is _no_ return statement at all (as opposed to some branches that don't have a return statement). But in TS this isn't a big deal, since the returned type will have `void`, which you can't do anything with.

@typescript-eslint/triage-team ?

**Comment by kirkwaiblinger** [comment]
To elaborate, `noImplicitReturns` essentially solves `.map()`, `.reduce()`, and friends.

TS itself solves `.sort()` and friends (by requiring a `number` to be returned)

And #10106 will solve `.some()`/`.every()`/, and `.find()` and friends.

So I don't see much remaining area of applicability for an extension rule, that's not already addressed somewhere else.
